File: packages/cli-node/src/commands/customize.ts
// spec-init customize --add/--remove/--persona: update spec.config.js (ground
// truth), re-derive CLAUDE.md's managed region, and add/remove integration files.

import { existsSync, readFileSync, readdirSync, rmSync, writeFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { UsageError } from '../args.js';
import type { CliOptions, Integration } from '../args.js';
import { resolvePayloadDir } from '../payload.js';
import { isManaged, readConfig, writeConfig } from '../config.js';
import { swapManagedRegions } from '../managed-region.js';
import { installIntegrations } from '../integrations.js';
import { applyIntegrations, maybeTransform, walkPayload } from './init.js';

/** Run `customize`. Requires exactly one of --add / --remove / --persona. */
export function runCustomize(opts: CliOptions): void {
  const projectRoot = resolve(process.cwd());
  if (!isManaged(projectRoot)) {
    throw new UsageError('customize must be run inside a scaffolded project (spec.config.js not found)');
  }

  const actions = [opts.addIntegration, opts.removeIntegration, opts.persona].filter(Boolean).length;
  if (actions === 0) throw new UsageError('customize requires --add, --remove, or --persona');
  if (actions > 1) {
    throw new UsageError('customize accepts one of --add / --remove / --persona per invocation');
  }

  const cfg = readConfig(projectRoot);
  const next = { persona: cfg.persona, integrations: [...cfg.integrations] };
  let target: Integration | null = null;
  let mode: 'add' | 'remove' | 'persona' = 'persona';

  if (opts.addIntegration) {
    mode = 'add';
    target = opts.addIntegration;
    if (!next.integrations.includes(target)) next.integrations.push(target);
  } else if (opts.removeIntegration) {
    mode = 'remove';
    target = opts.removeIntegration;
    next.integrations = next.integrations.filter((n) => n !== target);
  } else if (opts.persona) {
    next.persona = opts.persona;
  }

  const payloadDir = resolvePayloadDir();

  if (opts.dryRun) {
    process.stdout.write(
      `[dry-run] would set persona=${next.persona ?? 'engineer'}, integrations=[${next.integrations.join(', ')}]\n`,
    );
    return;
  }

  // 1. Persist config (ground truth).
  writeConfig(projectRoot, next);

  // 2. Re-derive CLAUDE.md's managed region for the new persona/integrations.
  const claudeOurs = join(projectRoot, 'CLAUDE.md');
  if (existsSync(claudeOurs)) {
    const theirs = maybeTransform(
      'CLAUDE.md',
      readFileSync(join(payloadDir, 'CLAUDE.md'), 'utf8'),
      next.persona,
      next.integrations,
    );
    const swapped = swapManagedRegions(readFileSync(claudeOurs, 'utf8'), theirs);
    if (swapped) writeFileSync(claudeOurs, swapped);
  }

  // 3. Integration files: add copies them in, remove deletes them.
  let touched = 0;
  if (mode === 'add' && target) {
    touched = applyIntegrations(payloadDir, projectRoot, [target]);
  } else if (mode === 'remove' && target) {
    touched = removeIntegrationFiles(payloadDir, projectRoot, target);
  }

  // 4. Auto-install a newly added integration (respects --no-install / non-TTY).
  if (mode === 'add' && target) {
    installIntegrations([target], { isTty: Boolean(process.stdin.isTTY), noInstall: opts.noInstall });
  }

  process.stdout.write(
    `[OK] customize: persona=${next.persona ?? 'engineer'}, integrations=[${next.integrations.join(', ')}] ` +
      `(${touched} integration file(s) changed)\n`,
  );
}

/** Delete an integration's files from the project and prune emptied dirs. */
function removeIntegrationFiles(payloadDir: string, projectRoot: string, target: Integration): number {
  const root = join(payloadDir, '_integrations', target);
  if (!existsSync(root)) return 0;
  let count = 0;
  const dirs = new Set<string>();
  for (const rel of walkPayload(root)) {
    const dest = join(projectRoot, rel);
    if (existsSync(dest)) {
      rmSync(dest);
      count += 1;
      dirs.add(dirname(rel));
    }
  }
  pruneEmpty(projectRoot, dirs);
  return count;
}

/** Remove now-empty directories that hosted an integration's files. */
function pruneEmpty(outDir: string, dirs: Set<string>): void {
  const ordered = [...dirs].sort((a, b) => b.length - a.length);
  for (const relDir of ordered) {
    let cur = relDir;
    while (cur && cur !== '.') {
      const abs = join(outDir, cur);
      if (!existsSync(abs)) break;
      if (readdirSync(abs).length > 0) break;
      rmSync(abs, { recursive: true });
      cur = dirname(cur);
    }
  }
}

File: packages/cli-python/src/spec_init/commands/customize.py
"""spec-init customize --add/--remove/--persona: update spec.config.js (ground
truth), re-derive CLAUDE.md's managed region, and add/remove integration files.
"""

from __future__ import annotations

import sys
from pathlib import Path

from ..args import CliOptions, Integration, UsageError
from ..config import ProjectConfig, is_managed, read_config, write_config
from ..integrations import install_integrations
from ..managed_region import swap_managed_regions
from ..payload import resolve_payload_dir
from .init import _maybe_transform, _walk_payload, apply_integrations


def run_customize(opts: CliOptions) -> None:
    """Update config, re-derive CLAUDE.md's managed region, add/remove integration files."""
    project_root = Path.cwd()
    if not is_managed(project_root):
        raise UsageError(
            "customize must be run inside a scaffolded project (spec.config.js not found)"
        )

    actions = sum(1 for a in (opts.add_integration, opts.remove_integration, opts.persona) if a)
    if actions == 0:
        raise UsageError("customize requires --add, --remove, or --persona")
    if actions > 1:
        raise UsageError("customize accepts one of --add / --remove / --persona per invocation")

    cfg = read_config(project_root)
    next_persona = cfg.persona
    next_integrations = list(cfg.integrations)
    target: Integration | None = None
    mode = "persona"

    if opts.add_integration:
        mode = "add"
        target = opts.add_integration
        if target not in next_integrations:
            next_integrations.append(target)
    elif opts.remove_integration:
        mode = "remove"
        target = opts.remove_integration
        next_integrations = [n for n in next_integrations if n != target]
    elif opts.persona:
        next_persona = opts.persona

    payload_dir = resolve_payload_dir()

    if opts.dry_run:
        sys.stdout.write(
            f"[dry-run] would set persona={next_persona or 'engineer'}, "
            f"integrations=[{', '.join(next_integrations)}]\n"
        )
        return

    # 1. Persist config (ground truth).
    write_config(
        project_root, ProjectConfig(persona=next_persona, integrations=tuple(next_integrations))
    )

    # 2. Re-derive CLAUDE.md's managed region for the new persona/integrations.
    claude = project_root / "CLAUDE.md"
    if claude.exists():
        theirs = _maybe_transform(
            "CLAUDE.md",
            (payload_dir / "CLAUDE.md").read_text(encoding="utf-8"),
            next_persona,
            tuple(next_integrations),
        )
        swapped = swap_managed_regions(claude.read_text(encoding="utf-8"), theirs)
        if swapped:
            claude.write_text(swapped, encoding="utf-8", newline="")

    # 3. Integration files: add copies them in, remove deletes them.
    touched = 0
    if mode == "add" and target:
        touched = apply_integrations(payload_dir, project_root, (target,))
    elif mode == "remove" and target:
        touched = _remove_integration_files(payload_dir, project_root, target)

    # 4. Auto-install a newly added integration (respects --no-install / non-TTY).
    if mode == "add" and target:
        install_integrations((target,), is_tty=sys.stdin.isatty(), no_install=opts.no_install)

    sys.stdout.write(
        f"[OK] customize: persona={next_persona or 'engineer'}, "
        f"integrations=[{', '.join(next_integrations)}] ({touched} integration file(s) changed)\n"
    )


def _remove_integration_files(payload_dir: Path, project_root: Path, target: Integration) -> int:
    """Delete an integration's files from the project and prune emptied dirs."""
    root = payload_dir / "_integrations" / target
    if not root.exists():
        return 0
    count = 0
    dirs: set[str] = set()
    for rel in _walk_payload(root):
        dest = project_root / rel
        if dest.exists():
            dest.unlink()
            count += 1
            dirs.add(str(Path(rel).parent).replace("\\", "/"))
    _prune_empty(project_root, dirs)
    return count


def _prune_empty(out_dir: Path, dirs: set[str]) -> None:
    """Remove now-empty directories that hosted an integration's files."""
    for rel in sorted(dirs, key=len, reverse=True):
        cur = rel
        while cur and cur != ".":
            abs_path = out_dir / cur
            if not abs_path.exists():
                break
            if any(abs_path.iterdir()):
                break
            abs_path.rmdir()
            cur = str(Path(cur).parent).replace("\\", "/")
